documentation/envConfiguration.md
# ⚙️ Environment Variables Configuration Explanation

This document provides an explanation of the environment variables for the configuration of the application.  
Ensure that you have set these environment variables appropriately in your deployment environment or in the `.env` file in the root directory of the project.

## 💾 Database

### MONGODB_URI (Required!)

- **Description:** The MongoDB URI to connect to the cluster (without the database name or any other parameters)
- **Example:**
  ```env
  MONGODB_URI=mongodb+srv://username:password@cluster0.1a2b3c4.mongodb.net/
  ```
- **Additional information:** For more information on the MongoDB Connection String URI Format, refer to the [MongoDB Connection Strings documentation](https://docs.mongodb.com/manual/reference/connection-string/).  
  When using Docker and local Redis, make sure you set the network and the URI properly.

### DB_NAME

- **Description:** The name of the database to use.
- **Default value:** `sample_mflix`.
- **Example:**
  ```env
  DB_NAME=mflix
  ```

### CONNECT_DB_TIMEOUT_MS

- **Description:** The maximum time to wait for a database connection to be established.
- **Default value:** `15000` (15 seconds).

### MAX_DB_POOL_SIZE

- **Description:** The maximum pool size of the database connections.
- **Default value:** `20`.
- **Additional information:** For more information, refer to [Connection Pooling in MongoDB](https://mongodb.com/blog/post/server-side-connections-are-coming-to-the-node-js-driver).

### MIN_DB_POOL_SIZE

- **Description:** The minimum pool size of the database connections.
- **Default value:** `10`.
- **Additional information:** For more information, refer to [Connection Pooling in MongoDB](https://mongodb.com/blog/post/server-side-connections-are-coming-to-the-node-js-driver).

### MAX_IDLE_TIME_DB_MS

- **Description:** The maximum time a connection can remain idle in the pool before being removed and closed.
- **Default value:** `60000` (1 minute).
- **Additional information:** For more information, refer to [Connection Pooling in MongoDB](https://mongodb.com/blog/post/server-side-connections-are-coming-to-the-node-js-driver).

### WRITE_CONCERN

- **Description:** The write concern level of the database.
- **Available choices:** `majority`, `0`, `1`, `2`.
- **Default value:** `majority`.
- **Additional information:** For more information, refer to [MongoDB Write Concern](https://mongodb.com/docs/manual/reference/write-concern/).

### WRITE_CONCERN_TIMEOUT

- **Description:** The write concern timeout of the database in milliseconds.
- **Default value:** `3000` (3 seconds).
- **Additional information:** For more information, refer to [MongoDB Write Concern](https://mongodb.com/docs/manual/reference/write-concern/).

### RECONNECTING_INTERVAL_DB_S

- **Description:** The interval in which the application will try to reconnect to the database if the connection fails (in seconds).
- **Default value:** `15`

## 🌐 Network

### PORT

- **Description:** The exposed port of the application.
- **Default value:** `3000`.
- **Additional information:** Make sure the port you choose is not already in use!

### ENABLE_LISTENING_TO_ALL_INTERFACES

- **Description:** Whether to listen to all network interfaces or not.
- **Default value:** `true`.
- **Additional information:** While [using Docker](dockerStart.md), this must be set to `true`.

## 🔐 Authentication

### JWT_SECRET

- **Description:** The secret of the JSON Web Token (JWT) used for authentication.
- **Default value:** A base64 encoded string of 256 random bytes.
- **Example:**
  ```env
  JWT_SECRET=OGUxNGE2M2hkM2YyMzAwMDE3NmQ4MmYxYz...
  ```

### JWT_EXPIRES_MINUTES

- **Description:** The expiration time of the JWT in minutes.
- **Default value:** `30`.

## 📝 Logging

### ENABLE_LOGGING

- **Description:** Whether to enable any logging or not.
- **Default value:** `true`.
- **Additional information:** If you disable logging, the following environment variables will be ignored.

### LOG_LEVEL

- **Description:** The log level of the application.
- **Available choices:** `fatal`, `error`, `warn`, `info`, `debug`, `trace`.
- **Default value:** `info`.

## 📖 Pagination

### DEFAULT_PAGE_SIZE

- **Description:** The default amount of entities to return per request.
- **Default value:** `15`.

### MAX_PAGE_SIZE

- **Description:** The maximum amount of entities to return per request.
- **Default value:** `50`.

## ⛔ Rate Limiting

### ENABLE_RATE_LIMITING

- **Description:** Whether to enable rate limiting or not.
- **Default value:** `false`.
- **Additional information:** If you disable rate limiting, the following environment variables will be ignored.

### RATE_LIMIT_WINDOW_MS

- **Description:** The time window in which the rate limit is applied (in milliseconds).
- **Default value:** `60000` (1 minute).

### RATE_LIMIT_MAX_REQUESTS

- **Description:** The maximum number of requests allowed in the time window.
- **Default value:** `30`.

### REDIS_URI

- **Description:** The Redis URI to use for the rate limiting.
- **Default value:** `redis://localhost:6379`.
- **Additional information:** When using Docker and local Redis, make sure you set the network and the URI properly.

## 🌠 Graceful Shutdown

### ENABLE_GRACEFUL_SHUTDOWN

- **Description:** Whether to enable graceful shutdown or not.
- **Default value:** `true`.
- **Additional information:** If you disable graceful shutdown, the following environment variables will be ignored.

### GRACEFUL_SHUTDOWN_TIMEOUT_LIMIT_MS

- **Description:** The maximum time to wait for the application to close gracefully (in milliseconds). If the application does not close gracefully within the timeout limit, it will be forcefully closed.
- **Default value:** `5000` (5 seconds).


documentation/envConfiguration.md.md
# ⚙️ Environment Variables Configuration Explanation

This document provides an explanation of the environment variables for the configuration of the application. The environment variables are used to customize various aspects of the application behavior. Ensure that you have set these environment variables appropriately in your deployment environment or in the `.env` file in the root directory of the application.

Below is the list of environment variables along with their descriptions and default values:

## Required Environment Variables

### MONGODB_URI

- The MongoDB URI to connect to the database (without the database name or any other parameters)
- Example:
  ```env
  MONGODB_URI=mongodb+srv://username:password@cluster0.1a2b3c4.mongodb.net/
  ```
  For more information on the MongoDB Connection String URI Format, refer to the [MongoDB Connection Strings documentation](https://docs.mongodb.com/manual/reference/connection-string/).

### DB_NAME

- The name of the database to use.
- Example:
  ```env
  DB_NAME=sample_mflix
  ```

## Optional Environment Variables

### LOG_LEVEL

- The log level of the application.
- Available choices: `error`, `warn`, `info`, `http`, `verbose`, `debug`, `silly`.
- Default value: `http`.

### PORT

- The exposed port of the application.  
  Make sure the port you choose is not already in use!
- Default value: `3000`.

### DEFAULT_PAGE_SIZE

- The default amount of entities to return per request.
- Default value: `15`.

### MAX_PAGE_SIZE

- The maximum amount of entities to return per request.
- Default value: `50`.

### CONNECT_DB_TIMEOUT_MS

- The maximum time to wait for a database connection to be established.
- Default value: `15000` (15 seconds).

### ENABLE_LISTENING_TO_ALL_INTERFACES

- Whether to listen to all network interfaces or not.  
  While using [Docker](dockerStart.md), this must be set to `true`.
- Default value: `true`.

### JWT_SECRET

- The secret of the JSON Web Token (JWT) used for authentication.
- Default value: A base64 encoded string of 256 random bytes.
- Example:
  ```env
  JWT_SECRET=OGUxNGE2M2hkM2YyMzAwMDE3NmQ4MmYxYz...
  ```

### JWT_EXPIRES_MINUTES

- The expiration time of the JWT in minutes.
- Default value: `30`.

### MAX_POOL_SIZE

- The maximum pool size of the database connections.  
  For more information, refer to [Connection Pooling in MongoDB](https://mongodb.com/blog/post/server-side-connections-are-coming-to-the-node-js-driver).
- Default value: `20`.

### MIN_POOL_SIZE

- The minimum pool size of the database connections.  
  For more information, refer to [Connection Pooling in MongoDB](https://mongodb.com/blog/post/server-side-connections-are-coming-to-the-node-js-driver).
- Default value: `10`.

### MAX_IDLE_TIME_MS

- The maximum time a connection can remain idle in the pool before being removed and closed.  
  For more information, refer to [Connection Pooling in MongoDB](https://mongodb.com/blog/post/server-side-connections-are-coming-to-the-node-js-driver).
- Default value: `60000` (1 minute).

### WRITE_CONCERN

- The write concern level of the database.  
  For more information, refer to [MongoDB Write Concern](https://mongodb.com/docs/manual/reference/write-concern/).

- Available choices: `majority`, `0`, `1`, `2`.
- Default value: `majority`.

### WRITE_CONCERN_TIMEOUT

- The write concern timeout of the database in milliseconds.  
  For more information, refer to [MongoDB Write Concern](https://mongodb.com/docs/manual/reference/write-concern/).
- Default value: `3000` (3 seconds).

### ENABLE_RECONNECTING

- Whether to enable reconnecting to the database if the connection fails.
- Default value: `true`.

### RECONNECTING_INTERVAL_MS

- The interval in which the application will try to reconnect to the database if the connection fails.
  If [`ENABLE_RECONNECTING`](#enable_reconnecting) is false, this property will be ignored.
- Default value: `15000` (15 seconds).

Please make sure to set these environment variables according to your specific deployment environment to ensure the proper functioning of the application.


src/setup/env.ts
import { config } from "dotenv";
import { randomBytes } from "crypto";
import { Level } from "pino";
import { cleanEnv, str, port, url, num, makeValidator, bool } from "envalid";

const getRandomJWTSecret = () => btoa(randomBytes(256).toString("hex"));

const availableLogLevel: readonly Level[] = [
  "fatal",
  "error",
  "warn",
  "info",
  "debug",
  "trace",
] as const;

config();

/**
 * The environment variables of the application.
 * For more info, see the @link[Environment Variables Documentation](../../documentation/envConfiguration.md).
 * @example
 * ```ts
 * console.log(env.PORT); // 3000
 * ```
 */
export const env = cleanEnv(process.env, {
  /**
   * The MongoDB URI to connect to the cluster.
   * @see[MongoDB Connection String URI Format](https://mongodb.com/docs/connections\trings/)
   * @example
   * ```env
   * MONGODB_URI=mongodb+srv://username:password@cluster0.1a2b3c4.mongodb.net/
   * ```
   */
  MONGODB_URI: url(),

  /**
   * The name of the DB to use.
   * @default "sample_mflix"
   */
  DB_NAME: str({ default: "sample_mflix" }),

  /**
   * Whether to enable logging or not.
   * @default true
   */
  ENABLE_LOGGING: bool({ default: true }),

  /**
   * The log level of the application.
   * @default "info"
   */
  LOG_LEVEL: str({
    choices: availableLogLevel,
    default: "info",
  }),

  /**
   * The exposed port of the application. Make sure there is no other service using this port.
   * @default 3000
   */
  PORT: port({ default: 3000 }),

  /**
   * Whether to enable graceful shutdown or not.
   * @default true
   */
  ENABLE_GRACEFUL_SHUTDOWN: bool({ default: true }),

  /**
   * The maximum time to wait for the application to close gracefully (in milliseconds).
   * @default 5000 (5 seconds)
   */
  GRACEFUL_SHUTDOWN_TIMEOUT_LIMIT_MS: num({ default: 5000 }),

  /**
   * The default amount of entities to return per request.
   * @default 15
   */
  DEFAULT_PAGE_SIZE: num({ default: 15 }),

  /**
   * The maximum amount of entities to return per request.
   * @default 50
   */
  MAX_PAGE_SIZE: num({ default: 50 }),

  /**
   * The max time to wait for a DB connection to be established.
   * @default 15000
   * @see[MongoDB Connection Options](https://mongodb.com/docs/drivers/node/current/fundam\entals/connection/connection-options/) for more info.
   * @see[Connection Pooling](https://mongodb.com/docs/drivers/node/current/fundam\entals/connection/pooling/) for more info.
   * @see[Connection Pooling in MongoDB](https://mongodb.com/blog/post/server-side-connections-are-coming-to-the-node-js-driver) for more info.
   */
  CONNECT_DB_TIMEOUT_MS: num({ default: 15000 }),

  /**
   * Whether to listening to all network interfaces or not.
   * @default true
   */
  ENABLE_LISTENING_TO_ALL_INTERFACES: bool({ default: true }),

  /**
   * The secret of the JWT.
   * @default "A base64 encoded string of 256 random bytes"
   * @example OGUxNGE2M2hkM2YyMzAwMDE3NmQ4MmYxYz...
   */
  JWT_SECRET: str({
    default: getRandomJWTSecret(),
  }),

  /**
   * The expiration time of the JWT in minutes.
   * @default 30
   */
  JWT_EXPIRES_MINUTES: num({ default: 30 }),

  /**
   * The max pool size of the DB.
   * @default 20
   * @see[Connection Pool Overview](https://www.mongodb.com/docs/manual/administration/connection-pool-overview/) for more info.
   */
  MAX_DB_POOL_SIZE: num({ default: 20 }),

  /**
   * The min pool size of the DB.
   * @default 10
   * @see[Connection Pool Overview](https://www.mongodb.com/docs/manual/administration/connection-pool-overview/) for more info.
   */
  MIN_DB_POOL_SIZE: num({ default: 10 }),

  /**
   * The maximum time a connection can remain idle in the pool before being removed and closed.
   * @default 30000
   * @see[Connection Pool Overview](https://www.mongodb.com/docs/manual/administration/connection-pool-overview/) for more info.
   */
  MAX_IDLE_TIME_DB_MS: num({ default: 30000 }),

  /**
   * The write concern level of the DB.
   * @default "majority"
   * @see[MongoDB Write Concern](https://mongodb.com/docs/manual/reference/write-concern/)
   */
  WRITE_CONCERN: makeValidator((input: string) => {
    if (input === "majority") return input;
    const num = Number(input);
    if (num === 0 || num === 1 || num === 2) return num;
    throw new Error("WRITE_CONCERN must be 0, 1, 2, or 'majority'!");
  })({
    default: "majority",
  }),

  /**
   * The write concern timeout of the DB.
   * @default 3000
   * @see[MongoDB Write Concern](https://mongodb.com/docs/manual/reference/write-concern/)
   */
  WRITE_CONCERN_TIMEOUT: num({
    default: 3000,
  }),

  /**
   * The interval in which the DB will try to reconnect to the DB if the connection is failed (in seconds).
   * @default 15
   */
  RECONNECTING_INTERVAL_DB_S: num({
    default: 15,
  }),

  /**
   * Whether to enable the rate limiter or not.
   * @default false
   */
  ENABLE_RATE_LIMITING: bool({ default: false }),

  /**
   * Redis URL to connect to for the rate limiter.
   *
   * @default "redis://localhost:6379"
   * @example ```env
   * REDIS_URL=rediss://ami:password@my-host:6379
   * ```
   */
  REDIS_URL: url({
    default: "redis://localhost:6379",
  }),

  /**
   * The interval in which the Redis will try to reconnect to the Redis if the connection is failed (in seconds).
   * @default 5
   */
  RECONNECTING_INTERVAL_REDIS_S: num({
    default: 5,
  }),

  /**
   * The window of the rate limiter in milliseconds.
   * @default 60000 (1 minute)
   */
  RATE_LIMIT_WINDOW_MS: num({
    default: 60000,
  }),

  /**
   * The max amount of requests per window.
   * @default 30
   */
  RATE_LIMIT_MAX_REQUESTS_PER_WINDOW: num({
    default: 30,
  }),
});

/**
 * The environment variables of the application.
 */
export type EnvironmentVariables = typeof env;
